### ISSUE TITLE
Deploy our own nbviewer for mybinder.org

### ISSUE THREAD

**Issue by choldgraf** [jupyterhub/team-compass#181]
What do people think about deploying our own nbviewer for mybinder.org? We've had a few conversations around curated or example repositories that show off different use-cases in the Binder ecosystem, it would be awesome if we could have a repository that nbviewer will automatically pull from and update (e.g. at `gallery.mybinder.org`).

One question for @parente is how difficult it'd be for us to run our own nbviewer as a service (perhaps this is something we could kubernetes-ify?) or if there's a different service that makes more sense?

What do folks think about this more generally?

**Comment by betatim** [comment]
Could you explain a bit more what we'd gain from having `nbviewer.mybinder.org` instead of using the nbviewer that is already running? Also what is "repo that nbviewer will automatically pull from"? Is that the gallery part of nbviewer.jupyter.org?

<img width="977" alt="Screen Shot 2019-07-01 at 21 31 25" src="https://user-images.githubusercontent.com/1448859/60461704-a2d52700-9c47-11e9-928b-d592c39dc7c1.png">

I think we should start with a gallery (statically generated, hosted on netlify or some such). That would fill our need and provide value to users without taking on the ops burden of having to maintain a nbviewer instance (which we get "for free" by linking to the existing one for our gallery).

**Comment by choldgraf** [comment]
@betatim linking out to nbviewer would work too, I was just thinking of it as a way to quickly get a gallery-like setup that points to repositories, which it seems nbviewer is already set up to do. (e.g. in the way that the page you linked to has a section for "programming languages", I'd like our Binder gallery to have different sections we have control over, such as "reproducibility" or "education" or "language-specific repositories" etc)

**Comment by betatim** [comment]
kk. Thanks for expanding your explanation. I think I'd rather take a moment to extract the HTML+CSS from the nbviewer page, put it in a jinja2 or Hugo or static-site-genrator-of-choice template, and feed it a yaml with repos to generate a gallery.

Operating a nbviewer instance sounds like a lot more work than using a hoster like netlify in terms of effort and things that can go wrong. If we do have spare resources to do that I think we should contribute them to the ops team of nbviewer.jupyter.org which is currently ~0 people (just a guess).

**Comment by minrk** [comment]
nbviewer is ~all fetching notebooks from hosting APIs and rendering them with nbviewer, with lots of caching logic to avoid hitting API rate limits.

The front page, on the other hand, is a pretty [basic page](https://github.com/jupyter/nbviewer/blob/master/nbviewer/templates/index.html) populated from [fixed, customizable data](https://github.com/jupyter/nbviewer/blob/master/nbviewer/frontpage.json). So if we want the list of links, I think we can copy the style of nbviewer's implementation. I'm pretty sure this can be a pretty basic basic jekyll/whatever site, even hosted on gh-pages without needing any running server or add any new services/accounts/etc.

If we want the source data to be more dynamic (e.g. top 10 repos in the last week), we can copy what nbviewer does for the index page more directly into binderhub and update the data source live. Or do it standalone with netlify/heroku/whatever.

On the other hand...I'd love an nbviewer helm chart to exist!

**Comment by parente** [comment]
> The front page, on the other hand, is a pretty basic page populated from fixed, customizable data. So if we want the list of links, I think we can copy the style of nbviewer's implementation. I'm pretty sure this can be a pretty basic basic jekyll/whatever site, even hosted on gh-pages without needing any running server or add any new services/accounts/etc.

An external gallery that redirects to nbviewer.jupyter.org URLs in the appropriate format would definitely be easier than hosting a new instance.

That said ...

> contribute them to the ops team of nbviewer.jupyter.org which is currently ~0 people (just a guess).

I think @minrk and I are the only ones who have deployed nbviewer to the Rackspace account where it's hosted in the past 2-3 years. I've only deployed new releases as people have submitted PRs, never really helped out with downtime like Min has.

> On the other hand...I'd love an nbviewer helm chart to exist!

👍 Moving to a more modern and standard config management system like helm would definitely help increase the pool of folks able and willing to help out with nbviewer ops. The current config management system is a mix of pyinvoke, docker machine, and docker, and predates k8s.

**Comment by choldgraf** [comment]
I wonder if "create a helm chart for nbviewer" would be a good project for an infra-oriented outreachy student or something? (cc @yuvipanda)

**Comment by consideRatio** [comment]
If a binderhub is deployed to access a private git-repo, it could perhaps allow nbviewer to show private stuff as well, as compared to the public one. Hmm..

I'm quite motivated to learn more and Helm charterize a nbviewer deployment, but I'm very low on available time and focus at the moment.

**Comment by minrk** [comment]
> I wonder if "create a helm chart for nbviewer" would be a good project for an infra-oriented outreachy student or something?

Yes, I think it would. It's an interesting and definitely useful and relatively small project.
